src/types.ts

```diff
@@ -400,3 +403,23 @@ export type Fetch = (
   input: RequestInfo | URL,
   init?: RequestInit | undefined,
 ) => Promise<Response>;
+
+export type RootOnUpdate = (
+  href: DOMString | null | undefined,
+  action: DOMString | null | undefined,
+  element: Element,
+  options: HvComponentOptions,
+  callbacks: OnUpdateCallbacks,
+) => void;
+
+export type OnUpdateCallbacks = {
+  clearElementError: () => void;
+  getNavigation: () => Navigation;
+  getOnUpdate: () => HvComponentOnUpdate;
+  getStateDoc: () => Document;
+  getDoc: () => Document;
+  getStateUrl: () => string;
+  registerPreload: (id: number, element: Element) => void;
+  setNeedsLoad: () => void;
+  setState: (state: object) => void;
```

Comment:
any ways we could type the state with a more explicit type?

Reply:
I started down that road, but realized that `hv-screen` and `hv-route` have different states (at this point). I can take another look at doing this as I think all of the values are optional in the state object.